Tests/SwiftFusionTests/Inference/MetropolisTests.swift

```diff
@@ -0,0 +1,287 @@
+    subscript(index: Int) -> Pose {
+      for (updatedIndex, updateValue) in overlay.updates {
+        if updatedIndex == index {
+          return values[index].updatedView(updateValue)
+        }
+      }
+      return values[index]
+    }
+
+    /// The indices of all the variables.
+    var indices: Array<Pose>.Indices {
+      return values.indices
+    }
+
+    /// An update to the variables.
+    struct Update: IndexedVariables {
+      let updates: [(Int, Pose.Update)]
+      var indices: LazyMapCollection<[(Int, Pose.Update)], Int> {
+        return updates.lazy.map { $0.0 }
+      }
+    }
+
+    /// An update overlaid on top of `values`.
+    ///
+    /// Reads take this update into account. This allows an `updatedView` implementation whose
+    /// performance is independent of `variableCount`.
+    var overlay: Update
+
+    /// Creates `Variables` from the given `values` and `overlay`.
+    init(values: [Pose], overlay: Update) {
+      self.values = values
+      self.overlay = overlay
+    }
+
+    func updatedView(_ update: Update) -> Variables {
+      return Variables(
+        values: self.values,
+        overlay: Update(updates: self.overlay.updates + update.updates)
+      )
+    }
+
+    mutating func update(_ update: Update) {
+      precondition(self.overlay.updates.count == 0)
+      for (updatedIndex, updateValue) in update.updates {
+        self.values[updatedIndex] = self.values[updatedIndex].updatedView(updateValue)
+      }
+    }
+  }
+
+  var variablesZero: Variables {
+    return Variables(Array(repeating: Pose(), count: variableCount))
+  }
+
+  func vertexId(of variable: Variables.Index) -> Graph.VertexId {
+    return Int32(variable)
+  }
+
+  func variableIndex(of vertex: Graph.VertexId) -> Variables.Index {
+    return Int(vertex)
+  }
+
+  // MARK: - Factor error function.
+
+  func error(of vertexId: Graph.VertexId, at point: Variables) -> Double {
+    let variableIndices =
+      graph.edges(from: vertexId).map { variableIndex(of: graph.destination(of: $0)) }
+    switch graph[vertex: vertexId] {
+    case .prior(prior: let prior):
+      assert(variableIndices.count == 1)
+      return prior.localCoordinate(point[variableIndices[0]]).squaredNorm
+    case .between(difference: let difference):
+      assert(variableIndices.count == 2)
+      let actualDifference = between(point[variableIndices[1]], point[variableIndices[0]])
+      return difference.localCoordinate(actualDifference).squaredNorm
+    case .none:
+      preconditionFailure("not a factor vertex")
+    }
+  }
+}
+
+extension Pose2: UpdatedViewable {
+  public typealias Update = Self
+  public func updatedView(_ update: Self) -> Self {
+    return update
+  }
+}
+
+/// An example proposal to use in the test.
+extension PoseSLAMFactorGraphForSmallProposals where Pose == Pose2 {
+  /// Selects a variable uniformly at random, and some of the "nearby" variables, and applies the
+  /// same random movement to all of them.
```

Comment:
So this is the random walk kernel Q

Reply:
(should have a question mark here, sorry :P )

Reply:
Could be? I don't know what that is.

I just came up with an arbitrary proposal that does something interested with the graph structure, to illustrate possibilities. It might be a pretty bad one. Or it might have a name?